Extend LPDDR5 to support BG mode (Interface power)

alexcani opened the pull request.

Commit:
Create BankArchitectureMode enum and parse bank architecture mode in LPDDR5 MemSpec

Commit:
Update command patterns for LPDDR5 depending on bank architecture

alexcani committed:
Add Bank Group bits on pattern encoding

alexcani committed:
LPDDR5 BG mode unit test

lsteiner-tukl commented on src/DRAMPower/DRAMPower/command/Pattern.h:
This file should be completed. Many cases in the switch are missing.

alexcani replied:
i agree, but maybe on a different task?

lsteiner-tukl replied:
sure

lsteiner-tukl commented on src/DRAMPower/DRAMPower/memspec/MemSpecLPDDR5.h:
Just to be consistent with the standard and DRAMSys I would prefer the following:
"enum class BankMode { M16B, MBG, M8B } bankMode;"

alexcani replied:
renamed!

lsteiner-tukl commented on src/DRAMPower/DRAMPower/memspec/MemSpecLPDDR5.cpp:
```diff
@@ -66,7 +66,35 @@ MemSpecLPDDR5::MemSpecLPDDR5(nlohmann::json &memspec)
+        std::cout << "Selected values:" << std::endl;
+        std::cout << "  - Number of banks: " << numberOfBanks << std::endl;
+        std::cout << "  - Number of bank groups: " << numberOfBankGroups << std::endl;
+        std::cout << "Valid values are 16|0 (16B mode), 16|4 (BG mode) or 8|0 (8B mode)" << std::endl;
```
Actually it should be 16|1, 16|4 or 8|1.

alexcani replied:
fixed

lsteiner-tukl commented on src/DRAMPower/DRAMPower/standards/lpddr5/LPDDR5.cpp:
```diff
@@ -47,58 +47,121 @@ namespace DRAMPower {
+        }
+        this->registerPattern<CmdType::PRE>(pre_pattern);
+
+        commandPattern_t prea_pattern = {
```
For PREA and REFA the bank address is a don't care, so we could also assign V.

alexcani replied:
done

lsteiner-tukl commented on src/DRAMPower/DRAMPower/standards/lpddr5/LPDDR5.cpp:
```diff
@@ -47,58 +47,121 @@ namespace DRAMPower {
+            refa_pattern[9] = BG0;
+        }
+        this->registerPattern<CmdType::REFA>(refa_pattern);
         this->registerPattern<CmdType::REFP2B>({
```
We should also add support for the REFP2B. From the interface perspective there is no difference to a REFPB, only the core power calculations might be different because two banks are refreshed at once.

alexcani replied:
REFP2B is not on the command table for lpddr5, but since it was already in the file when i started working on it I never removed. maybe I should just take this out? it doesn't seem to be a feature of lpddr5

lsteiner-tukl replied:
In LPDDR5 it's not called REFP2B, but if you have 16 banks or bank groups 2 banks are refreshed at the same time. So it's basically identical to a REFP2B, and within DRAMSys it's also handled as a REFP2B.

alexcani replied:
I read the standard a bit and i replicated the REFB for REFP2B if we're in 16B or BG mode. Can you confirm?

alexcani committed:
PR suggestions

Renamed bank architecture modes to match DRAMSys names and fixed error message
PREA and REFA bank address patterns are now a don't care (V)

alexcani committed:
Enable REFP2B

IESE-T3 commented:
Looks good to me!
